### PR TITLE
feat(aapd-402): add date component and community levy adopted date qu…

### PR DESCRIPTION
**Pull request by edpackard-work**
…estion

## Ticket Number

https://pins-ds.atlassian.net/browse/AAPD-402

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [ ] My commit history in this PR is linear
- [ ] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/dynamic-forms/validator/date-validator.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+				}),
+			body(yearInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (req.body[dayInput] && req.body[monthInput]) return this.noYearErrorMessage;
+				}),
+
+			// check date values entered are within valid ranges
+			body(dayInput)
+				.isInt({ min: 1, max: 31 })
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/validator/date-validator.js:94]: I'd consider a utility method like so
`var isValid = require('date-fns/isValid')
isValidWrapper(year = 2000, month = 1, day = 1)
{
isValid(new Date(year, month, 1)) : true ? throw InvalidMonthError("Month is invalid month " + month);
}
// Then we can do, which will work on any day, we can do the same to check the month and year on their own above
if (isValidWrapper(day = value) {
						const yearNum = parseInt(req.body[yearInput], 10);
						return yearNum && isLeapYear(new Date(yearNum, 1, 1)) && value === 29;
					}

`

File: packages/forms-web-app/src/dynamic-forms/validator/date-validator.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+					}
+
+					return true;
+				})
+				.withMessage(this.invalidDateErrorMessage),
+			body(monthInput).isInt({ min: 1, max: 12 }).withMessage(this.invalidDateErrorMessage),
+			body(yearInput).isInt({ min: 1000, max: 9999 }).withMessage(this.invalidYearErrorMessage),
+
+			//check date is not in the future
+			body(dayInput).custom((value, { req }) => {
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/validator/date-validator.js:114]: We might need to be able to configure if the validation is a check on the future or the past, I think there is another question where it checks the date is allowed to be in the past but not the future

File: packages/forms-web-app/src/dynamic-forms/validator/date-validator.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+				const day = `0${value}`.slice(-2);
+				const month = `0${req.body[monthInput]}`.slice(-2);
+				const year = req.body[yearInput];
+
+				const dateInput = `${year}-${month}-${day}`;
+				const today = endOfDay(new Date());
+
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/validator/date-validator.js:121]: Neat

File: packages/forms-web-app/src/dynamic-forms/validator/date-validator.js (hunk 2 above)
Comment (edpackard-work) [reply packages/forms-web-app/src/dynamic-forms/validator/date-validator.js:118]: as discussed, we do not need to implement this functionality for this ticket - the existing functionality can be adapted in a future ticket when requirements demand.

**Review by chrishawldev — APPROVED** [review]

**Pull request merged**